### PR TITLE
Add scenerios to downgrade/upgrade virtio drivers

### PR DESCRIPTION
**Pull request by suqinhuang**
Signed-off-by: Suqin Huang <shuang@redhat.com>

### CODE REVIEW

File: qemu/tests/driver_update.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+    session = vm.wait_for_login()
+    driver_name = params.get("driver_name")
+    driver_info = get_driver_from_cdrom(driver_name, session)
+    driver_path = driver_info[0]
+    driver_to_install = driver_info[1]
+
+    error.context("Update %s to version %s" % \
```
Comment (xutian) [inline qemu/tests/driver_update.py]: no need the '\'

File: qemu/tests/driver_update.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        s, output = session.cmd_status_output(driver_ver_cmd)
+        if s:
+            raise error.TestFail(output)
+        driver_version = output.split()[-1].split(".")[-1]
+        return int(driver_version)
+
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+    session = vm.wait_for_login()
+    driver_name = params.get("driver_name")
```
Comment (xutian) [inline qemu/tests/driver_update.py]: driver_name used in below, so better to use params["driver_name"], if not set test will failed ASAP.

File: qemu/tests/driver_update.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        guest_n = guest_list[guest_name]
+        virtio_win_disk = utils_misc.get_win_disk_vol(session, device_key) 
+        if not virtio_win_disk:
+            raise error.Error("Didn't fine virtio-win iso in guest")
+
+        driver_path = "%s:\%s\%s" % (virtio_win_disk, driver_name, guest_n)
+        cmd = "dir /B %s\*.inf" % driver_path
+        status, output = session.cmd_status_output(cmd) 
+        if status:
+            raise error.TestError("%s" % output)
```
Comment (xutian) [inline qemu/tests/driver_update.py]: output is string, so no need to convert it here

File: qemu/tests/driver_update.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+        guest_name_alias = params.get("guest_alias")
+        guest_name = params.get("guest_name")
+        guest_list = {}
+        for i in guest_name_alias.split(","):
+            guest, guest_alias = i.split(":")
+            guest_list[guest] = guest_alias
+        guest_n = guest_list[guest_name]
+        virtio_win_disk = utils_misc.get_win_disk_vol(session, device_key) 
+        if not virtio_win_disk:
+            raise error.Error("Didn't fine virtio-win iso in guest")
```
Comment (xutian) [inline qemu/tests/driver_update.py]: what's mean ?

File: qemu/tests/driver_update.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+        """
+        error.context("Copy installation log from guest to host", logging.info)
+        debug_files = "C:\Windows\DPINST.log,C:\Windows\inf\setupapi.dev.log"
+        driver_logs = params.get("driver_debug_file", debug_files).split(",")
+        result_dir = test.resultsdir
+
+        for log_file in driver_logs:
+            logging.info("Copy debug file %s from guest to host" % log_file)
+            try:
+                vm.copy_files_from(log_file, result_dir)
```
Comment (xutian) [inline qemu/tests/driver_update.py]: why not use test.resultsdir directly?

File: qemu/tests/driver_update.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+            except FileTransferNotFoundError:
+                logging.error("Didn't find file %s in guest" % log_file)
```
Comment (xutian) [inline qemu/tests/driver_update.py]: Can't retrieve loging file (%s) from guest.

File: qemu/tests/driver_update.py [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+ 
+        :param driver_name: name of the driver
+        return: driver path and driver version
+        """
+        error.context("Get driver driver path and driver version"
+                       " from virtio-win iso", logging.info)
+        device_key = params.get("device_key", "VolumeName like '%virtio-win%'")
+        guest_name_alias = params.get("guest_alias")
+        guest_name = params.get("guest_name")
+        guest_list = {}
```
Comment (xutian) [inline qemu/tests/driver_update.py]: guest_list = params.get('guest_alias')
guest_list = dict([x.split(':') for x in guest_list.split(',')])

File: qemu/tests/driver_update.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+        driver_path = "%s:\%s\%s" % (virtio_win_disk, driver_name, guest_n)
+        cmd = "dir /B %s\*.inf" % driver_path
+        status, output = session.cmd_status_output(cmd) 
+        if status:
+            raise error.TestError("%s" % output)
+        if not output:
+            raise error.TestError("Didn't find files in %s, pls check"
+                        "if the driver folder is existing" % driver_path)
+        inf_file = output.split()[0]
+        inf_file = "%s\%s" % (driver_path, inf_file)
```
Comment (xutian) [inline qemu/tests/driver_update.py]: inf_file = "%s\%s" % (driver_path, output.split()[0])

File: qemu/tests/driver_update.py [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+
+        logging.info("Get driver version from %s" % inf_file)
+        output = session.cmd("type %s" % inf_file)
+        ver_pattern = "DriverVer=(.*)"
+        version = re.findall(ver_pattern, output)[0]
+        driver_ver = int(version.split(".")[-1])
```
Comment (xutian) [inline qemu/tests/driver_update.py]: I think no need to convert version to int then do compare.  compare version string is OK.

**Comment by xutian** [comment]
deps on https://github.com/avocado-framework/avocado-vt/pull/334

File: qemu/tests/driver_update.py [hunk 10]
```diff
@@ -0,0 +1,148 @@
+import time
```
Comment (xutian) [inline qemu/tests/driver_update.py]: Please adjust order of import modules.   

import os 
import time
import re
import logging

File: qemu/tests/cfg/driver_update.cfg [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+            image_size_stg = 4G
+            force_create_image_stg = yes
+        - with_vioscsi:
+            driver_name = vioscsi
+            device_name = "Red Hat VirtIO SCSI pass-through controller"
+            drive_format_image1 = ide
+            drive_format_stg = scsi-hd
+            images += " stg"
+            image_name_stg = "images/stg"
+            image_size_stg = 4G
```
Comment (spcui) [inline qemu/tests/cfg/driver_update.cfg]: lose this line:
force_create_image_stg = yes

File: qemu/tests/driver_update.py [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+import logging
+import os
+
+from autotest.client.shared import error
+from autotest.client.shared import utils
+
+from virttest import utils_test
+from virttest import utils_misc
+from virttest import data_dir
+from virttest import env_process
```
Comment (spcui) [inline qemu/tests/driver_update.py]: unused modules: time, os, utils, data_dir, env_process

File: qemu/tests/driver_update.py (hunk 6 above)
Comment (spcui) [inline qemu/tests/driver_update.py]: undefined name 'FileTransferNotFoundError'

File: qemu/tests/driver_update.py (hunk 4 above)
Comment (suqinhuang) [inline qemu/tests/driver_update.py]: if virtio_win iso is not attached or recognited in guest

File: qemu/tests/driver_update.py (hunk 3 above)
Comment (suqinhuang) [inline qemu/tests/driver_update.py]: there may be one or more blank space after version in .inf file， so convert it to int
e.g "8600   " != "8600"

File: qemu/tests/driver_update.py (hunk 9 above)
Comment (suqinhuang) [inline qemu/tests/driver_update.py]: There may be one or more blank space after version
e.g "8600       " != "8600"

**Comment by CongSmile** [comment]
1. Miss step "enable driver verifier".
2. Only reinstall for the case? seems miss downgrade scenario.

File: qemu/tests/driver_update.py (hunk 9 above)
Comment (spcui) [inline qemu/tests/driver_update.py]: strip() can do it.

**Pull request closed without merging**